### DISCUSSION TITLE
How can I automate interactions with the gum binary (gum choose, gum input) in a BASH or other shell environment?

### DISCUSSION THREAD

**Discussion by bemidji3** [charmbracelet/gum#755] [Q&A]
I am working on a BASH script that uses gum to prompt the user for input and then generates a java project based on that input.  I am trying to automate the testing of my script and I am wondering how I can pass the arrow keys and the enter key via standard in to the gum binary in a shell environment.  Is this something that is supported right now?  

```
echo "text\n" | gum input
```
^this doesn't work as of right now nor does the use of \r or the encoded value for the enter key.

**Comment by wellkilo** [comment]
Piping bytes to stdin is not a reliable way to drive Gum's interactive widgets. `gum input` and `gum choose` first inspect/read piped stdin as **data** (initial value or choices), then start a Bubble Tea program for keyboard events. A pipe is not a terminal/PTY, and it may already be consumed before the interactive loop starts.

For an end-to-end CLI test, use a pseudo-terminal driver such as `expect`:

```tcl
#!/usr/bin/expect -f
set timeout 5
spawn gum input --prompt "Name: "
send -- "Alice\r"
expect eof
```

For `gum choose`:

```tcl
#!/usr/bin/expect -f
set timeout 5
spawn gum choose Alpha Beta Gamma
send -- "\033[B"  ;# Down arrow
send -- "\r"       ;# Enter
expect eof
```

Run with `expect test-gum.exp`. If you need to assert the selected value without terminal escape sequences, capture stdout separately; Gum renders the interactive UI to stderr and prints the submitted result to stdout. The current commands explicitly construct Bubble Tea with `tea.WithOutput(os.Stderr)` and print the final value afterwards: [`input`](https://github.com/charmbracelet/gum/blob/7179388031ae67d7f538d001be87d931f1cf5e28/input/command.go#L59-L78), [`choose`](https://github.com/charmbracelet/gum/blob/7179388031ae67d7f538d001be87d931f1cf5e28/choose/command.go#L140-L175).

For most shell-unit tests, an even more robust design is to separate “ask the user” from “generate the project”: test the generator with a normal argument/env var, and keep one small PTY smoke test for Gum interaction.

If this clarifies why `echo` cannot emulate the keypresses, please mark the reply as accepted.
